packages/siwe-parser/lib/abnf.ts

```diff
@@ -1,366 +1,162 @@
+    query: string | undefined;
+    fragment: string | undefined;
+  };
+
+  // For debugging: if doTrace = true apg-js will trace the parse tree
+  // and display it on an HTML page.
+  constructor(msg: string, doTrace = false) {
+    const grammarObj = new Grammar();
+    const parser = new apgLib.parser();
+    parser.callbacks["sign-in-with-ethereum"] = cb.signInWithEtherium;
+    parser.callbacks["oscheme"] = cb.oscheme;
+    parser.callbacks["domain"] = cb.domain;
+    parser.callbacks["LF"] = cb.lineno;
+    parser.callbacks["ex-title"] = cb.exTitle;
+    parser.callbacks["nb-title"] = cb.nbTitle;
+    parser.callbacks["ri-title"] = cb.riTitle;
+    parser.callbacks["re-title"] = cb.reTitle;
+    parser.callbacks["address"] = cb.address;
+    parser.callbacks["statement"] = cb.statement;
+    parser.callbacks["version"] = cb.version;
+    parser.callbacks["chain-id"] = cb.chainId;
+    parser.callbacks["nonce"] = cb.nonce;
+    parser.callbacks["issued-at"] = cb.issuedAt;
+    parser.callbacks["expiration-time"] = cb.expirationTime;
+    parser.callbacks["not-before"] = cb.notBefore;
+    parser.callbacks["request-id"] = cb.requestId;
+    parser.callbacks["uri"] = cb.uri;
+    parser.callbacks["uri-r"] = cb.uriR;
+    parser.callbacks["resource"] = cb.resource;
+    parser.callbacks["scheme"] = cb.scheme;
+    parser.callbacks["userinfo-at"] = cb.userinfo;
+    parser.callbacks["host"] = cb.host;
+    parser.callbacks["IP-literal"] = cb.ipLiteral;
+    parser.callbacks["port"] = cb.port;
+    parser.callbacks["path-abempty"] = cb.pathAbempty;
+    parser.callbacks["path-absolute"] = cb.pathAbsolute;
+    parser.callbacks["path-rootless"] = cb.pathRootless;
+    parser.callbacks["path-empty"] = cb.pathEmpty;
+    parser.callbacks["query"] = cb.query;
+    parser.callbacks["fragment"] = cb.fragment;
+    parser.callbacks["IPv4address"] = cb.ipv4;
+    parser.callbacks["nodcolon"] = cb.nodcolon;
+    parser.callbacks["dcolon"] = cb.dcolon;
+    parser.callbacks["h16"] = cb.h16;
+    parser.callbacks["h16c"] = cb.h16;
+    parser.callbacks["h16n"] = cb.h16;
+    parser.callbacks["h16cn"] = cb.h16;
+    parser.callbacks["dec-octet"] = cb.decOctet;
+    parser.callbacks["dec-digit"] = cb.decDigit;
+
+    // initialize parsed elements
+    const elements = {
+      errors: [],
+      lineno: 1,
+      scheme: undefined,
+      domain: undefined,
+      address: undefined,
+      statement: undefined,
+      uri: undefined,
+      version: undefined,
+      chainId: undefined,
+      nonce: undefined,
+      issuedAt: undefined,
+      expirationTime: undefined,
+      notBefore: undefined,
+      requestId: undefined,
+      resources: undefined,
+      uriElements: {
+        scheme: undefined,
+        userinfo: undefined,
+        host: undefined,
+        port: undefined,
+        path: undefined,
+        query: undefined,
+        fragment: undefined,
+      },
+    };
+    // uncomment for tracing
+    // if (doTrace) {
```

Comment:
Is there a way to automatically enable this when needed without having to uncomment code?

Reply:
Yes. I can make that happen. It will come in handy if you find yourself scratching your head over a message that you think is good but fails. You can usually spot the problem quickly in the trace. The trace will show up in a new folder `output/trace.html`.

Reply:
I didn't realise it was going to write to a file and make use of node dependencies -- isn't there a way to just output logs? It seems odd to me that a library would output files

Reply:
I used the trace extensively in my development and testing. However, the reason it was commented out in the first place was that I assumed that you wouldn’t want node dependencies in the production code. But if you want tracing as an option, node dependencies is the price to pay. 

For now I’m commenting the tracing code out again. However, I have a model for printing the trace to the console in real time in `apg-lite`. I’m not sure how much work it will be but I’ll take a look at moving that technique as an option into `apg-js`. I should have it fixed or will have given up on it by mid next week..